### ISSUE TITLE
Orphaned processes when using Singularity and trace

### ISSUE THREAD

**Issue by krokicki** [nextflow-io/nextflow#3606]
### Expected behavior and actual behavior

All descendant processes should be cleaned up when a Nextflow pipeline is interrupted or cancelled. This works in most cases, but when `singularity` and `trace` are both enabled the command.run file produces process trees which are not cleaned up when the pipeline is interrupted (e.g. using CTRL-C). 

### Steps to reproduce the problem

nextflow.config:
```
singularity.enabled = true
trace.enabled = true
```

test.nf
```
#!/usr/bin/env nextflow
nextflow.enable.dsl = 2
process test {
    container 'biocontainers/biocontainers:latest'
    shell:
    '''
    sleep 300
    '''
}
workflow {
    test()
}
```

### Program output 

When trace is not enabled, the process tree like this:

```
31887 ?        Ss     0:00  \_ sshd: rokickik [priv]
31892 ?        S      0:01  |   \_ sshd: rokickik@pts/7
31893 pts/7    Ss     0:07  |       \_ -bash
17285 pts/7    Sl+    0:22  |           \_ /usr/lib/jvm/zulu-17.0.3/bin/java (…) nextflow.cli.Launcher run test.nf
17423 pts/7    S+     0:00  |               \_ /bin/bash -ue .command.run
17439 pts/7    S+     0:00  |                   \_ tee .command.out
17440 pts/7    S+     0:00  |                   \_ tee .command.err
17441 pts/7    S+     0:00  |                   \_ /bin/bash -ue .command.run
17442 pts/7    Sl+    0:00  |                       \_ Singularity runtime parent
17464 pts/7    S+     0:00  |                           \_ /bin/bash -ue /groups/scicompsoft/home/rokickik/dev/test-ctrlc/work/21/7ab499f9054318ba49e7755b76551e/.command.sh
17491 pts/7    S+     0:00  |                               \_ sleep 300
```

When you CTRL-C the pipeline, it cleans up all of these processes via the `on_term` handler (thanks to [#1561](https://github.com/nextflow-io/nextflow/issues/1561)).

However, once you enable trace (or timeline/report/etc), the process tree looks like this:

```
31887 ?        Ss     0:00  \_ sshd: rokickik [priv]
31892 ?        S      0:01  |   \_ sshd: rokickik@pts/7
31893 pts/7    Ss     0:07  |       \_ -bash
14463 pts/7    Sl+    0:21  |           \_ /usr/lib/jvm/zulu-17.0.3/bin/java (…) nextflow.cli.Launcher run test.nf
14593 pts/7    S+     0:00  |               \_ /bin/bash -ue .command.run
14607 pts/7    S+     0:00  |                   \_ tee .command.out
14608 pts/7    S+     0:00  |                   \_ tee .command.err
14609 pts/7    S+     0:00  |                   \_ /bin/bash -ue .command.run
14610 pts/7    Sl+    0:00  |                       \_ Singularity runtime parent
14636 pts/7    S+     0:00  |                           \_ /bin/bash /groups/scicompsoft/home/rokickik/dev/test-ctrlc/work/46/6f46fb5069b1dbba55a9564a476bbe/.command.run nxf_trace
14670 pts/7    S+     0:00  |                               \_ /bin/bash -ue /groups/scicompsoft/home/rokickik/dev/test-ctrlc/work/46/6f46fb5069b1dbba55a9564a476bbe/.command.sh
14676 pts/7    S+     0:00  |                               |   \_ sleep 300
14672 pts/7    S+     0:00  |                               \_ /bin/bash /groups/scicompsoft/home/rokickik/dev/test-ctrlc/work/46/6f46fb5069b1dbba55a9564a476bbe/.command.run nxf_trace
```

Upon interruption with CTRL-C, the `on_term` handler calls kill on the parent process, but this leaves the user process as an orphan:
```
14670 pts/7    S      0:00 /bin/bash -ue /groups/scicompsoft/home/rokickik/dev/test-ctrlc/work/46/6f46fb5069b1dbba55a9564a476bbe/.command.sh
14676 pts/7    S      0:00  \_ sleep 300
```

### Environment 

* Nextflow version: 22.10.6.5843
* Java version: 17.0.3
* Operating system: Scientific Linux 7.8
* Bash version: 4.2.46(2)-release
* Bash version (inside container): 4.3.48(1)-release

### Additional context

The sleep is a simple test, but I have tested with other long running processes (e.g. `du -sh / > /dev/null 2>&1`) and they have the same behavior. 

With truly long running processes as with a transient Spark cluster, this can leave behind orphans that remain indefinitely. Usually I eventually run out of ports and need to clean them up manually.

**Comment by krokicki** [comment]
After messing around with the generated scripts, the only way to can get everything to work reliably was by adding an additional trap to the nxf_trace_linux function, e.g.

```
spid=
cleanup() {
    [[ $spid ]] && kill "$spid"
}
nxf_trace_linux() {
...
local task=$!
spid=$task
trap cleanup EXIT
...
wait $task
spid=
```

And of course I had to do the same thing in my process script for any long running background processes.

**Comment by krokicki** [comment]
Actually, this is a better script for testing this issue:
```
#!/usr/bin/env nextflow
nextflow.enable.dsl = 2
process test {
    container 'biocontainers/biocontainers:latest'
    shell:
    '''
    wpid=
    trap '[[ $wpid ]] && kill "$wpid"' EXIT
    sleep 300 & wpid=$!
    wait $wpid
    wpid=
    '''
}
workflow {
    test()
}
```

The reason is that [sleep is not interruptible, but wait is](http://mywiki.wooledge.org/SignalTrap#When_is_the_signal_handled.3F). The outcome is currently the same, but once Nextflow is fixed this script should work (all processes will be cleaned up) but the original one I posted will still have an orphan sleep process.

**Comment by bentsherman** [comment]
Hey Konrad, I drafted a PR with your suggested fix: #3620.

Please feel free to build the branch locally and test it with your setup, otherwise I will test it more thoroughly when I have some time.

**Comment by krokicki** [comment]
Thanks @bentsherman! I don't have time to figure out how to build Nextflow from a branch right now, but the code looks good to me.

**Comment by bentsherman** [comment]
No worries, I will get to it. But because I can't help myself, here is a quickstart...

```bash
# build
git clone -b 3606-singularity-trace-cleanup-child-procs git@github.com:nextflow-io/nextflow.git
cd nextflow
make compile

# test
../launch.sh run main.nf ...
```

Java is the only external dependency, so if you can run Nextflow then you can also build it 😄

**Comment by krokicki** [comment]
That's very helpful, thank you. And great to know that it is so easy to build 😄 

It works now when there is still a process running when you interrupt it, but when the user process terminates normally I'm getting this error:
```
.command.run: line 150: trap_pid: unbound variable
```

Adding a `set +u` to the start of cleanup() fixes that, but then I get this output for some reason:

```
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
```

When I rewrite it as follows, everything inexplicably works 🤷 
```
    cleanup() {
        set +u
        if [[ "$trap_pid" ]]; then
            kill $trap_pid 2>/dev/null
        fi
        set -u
    }
```

**Comment by bentsherman** [comment]
I was seeing that error in the CI test as well. I'd like to figure out the cleanest way to implement this hook in Bash ...

**Comment by bentsherman** [comment]
Hi Konrad, I'm going to try to replicate this issue on my machine. Which version of Singularity are you using?

**Comment by krokicki** [comment]
@bentsherman 3.8.7-1

**Comment by bentsherman** [comment]
Okay, I was able to reproduce the issue for both examples. When I swapped the biocontainers image with `'quay.io/nextflow/bash'`, the wait example did not leave any orphans. So that makes me think that the issue only exists for certain containers.

The main difference that I see between the two containers is the Bash version:
```console
$ singularity shell  work/singularity/biocontainers-biocontainers-latest.img 
Singularity> bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Singularity> exit
$ singularity shell work/singularity/quay.io-nextflow-bash.img 
Singularity> bash --version
GNU bash, version 4.4.12(1)-release (x86_64-openwrt-linux-gnu)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Singularity> exit
```

Maybe there is some difference between Bash 4.3 and 4.4 in how child processes are handled with traps? Seems unlikely.

Have you observed this behavior with other container images as well? Can you try with `'quay.io/nextflow/bash'` and let me know if it works on your Spark cluster?

**Comment by pditommaso** [comment]
I was giving to this. Having this process tree 

```
$ ps fx
    PID TTY      STAT   TIME COMMAND
1884936 ?        S      0:00 sshd: paolo@pts/1
1884937 pts/1    Ss     0:00  \_ -bash
1885529 pts/1    R+     0:00      \_ ps fx
1885202 pts/1    Sl     0:10 /usr/bin/java --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-ope
1885235 pts/1    S      0:00  \_ /bin/bash -ue .command.run
1885248 pts/1    S      0:00      \_ tee .command.out
1885249 pts/1    S      0:00      \_ tee .command.err
1885250 pts/1    S      0:00      \_ /bin/bash -ue .command.run
1885251 pts/1    Sl     0:00          \_ Singularity runtime parent
1885266 pts/1    S      0:00              \_ /bin/bash /home/paolo/test/work/54/7f5e95eb1334e743f3cb48897da5c8/.command.run nxf_trace
1885306 pts/1    S      0:00                  \_ /bin/bash -ue /home/paolo/test/work/54/7f5e95eb1334e743f3cb48897da5c8/.command.sh
1885312 pts/1    S      0:00                  |   \_ sleep 300
1885308 pts/1    S      0:00                  \_ /bin/bash /home/paolo/test/work/54/7f5e95eb1334e743f3cb48897da5c8/.command.run nxf_trace

```


trying to kill the root process (1885235) using the same code snippet using bt NF, it shows the children are all killed properly 

```
kill 1885235
kill 1885248
kill 1885249
kill 1885250
kill 1885251
kill 1885266
kill 1885306
kill 1885312
kill 1885308
```

**Comment by pditommaso** [comment]
Pushed this 477c65d0. tho still not 100% sure it solves the problem

**Comment by pditommaso** [comment]
I've double-checked this. It looks solved by the above commit

**Comment by krokicki** [comment]
Hi @pditommaso, I tested with [v23.03.0-edge](https://github.com/nextflow-io/nextflow/releases/tag/v23.03.0-edge) which includes your fix, but it doesn't seem to completely resolve the issue I'm seeing.

I did some testing and I think what's missing now is just a trap in the nxf_trace() function, something like what's there in nxf_main, e.g.:

```
nxf_trace() {
    trap on_term TERM INT
```

When I make this change, everything gets cleaned up correctly, in the tests above, and in my Spark workers.

Here's a PR: https://github.com/nextflow-io/nextflow/pull/3803

**Comment by krokicki** [comment]
Well, please ignore the PR, since it seems I'm not up to speed on the commit process for this project. But it's a one-line fix, and you probably want to test it yourself anyway.

**Comment by krokicki** [comment]
This issue was fully resolved in 23.04.0, see https://github.com/nextflow-io/nextflow/pull/3803#issuecomment-1492867161
